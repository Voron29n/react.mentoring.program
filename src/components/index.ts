export * from './common/Button';
export * from './common/Logo';

export * from './containers/Header';
export * from './containers/Lightbox';
export * from './containers/MovieForm';
export * from './containers/MovieForm/MovieFormView';
export * from './containers/MovieList';
export * from './containers/MovieList/MovieListView';

export * from './service/DeleteMovie';
export * from './service/EditMovie';
export * from './service/AddMovie';
export * from './service/SearchMovie';

export * from './single/Confirm';
export * from './single/Congratulations';
export * from './single/DropDown';
export * from './single/DropDown/DropDownItem';
export * from './single/Footer';
export * from './single/FilterBar';
export * from './single/GenreBar';
export * from './single/GenreBar/GenreBarItem';
export * from './single/GenreCheckmarks';
export * from './single/MovieDetail';
export * from './single/MovieItem/MovieView';
export * from './single/MovieItem';
export * from './single/MovieInput';
export * from './single/MovieInput/MovieInputView/index';
export * from './single/SearchBar';
export * from './single/SortBar';
export * from './single/Spinner';
export * from './single/Redirect';

export * from './error/ErrorBoundary';
export * from './error/FetchErrorMessage';
export * from './error/UnknownPage';
