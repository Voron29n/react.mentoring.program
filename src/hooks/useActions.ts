import { useDispatch } from 'react-redux';
import { bindActionCreators } from 'redux';
import { MoviesActionCreators } from 'store/actionCreators';

export const useActions = () => {
  const dispatch = useDispatch();
  return bindActionCreators(MoviesActionCreators, dispatch);
};
