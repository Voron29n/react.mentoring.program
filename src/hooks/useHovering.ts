import { RefObject, useEffect, useState } from 'react';

export const useHovering = (ref: RefObject<HTMLDivElement>) => {
  const [isHovering, setHovering] = useState<boolean>(false);

  const on: () => void = () => setHovering(true);
  const off: () => void = () => setHovering(false);

  useEffect(() => {
    if (!ref.current) {
      return;
    }

    const node = ref.current;

    node.addEventListener('mouseenter', on);
    node.addEventListener('mousemove', on);
    node.addEventListener('mouseleave', off);

    return () => {
      node.removeEventListener('mouseenter', on);
      node.removeEventListener('mousemove', on);
      node.removeEventListener('mouseleave', off);
    };
  }, [ref]);

  return isHovering;
};
