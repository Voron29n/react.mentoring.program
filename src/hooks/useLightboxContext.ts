import { useContext } from 'react';
import { LightboxContext } from 'context';

export const useLightboxContext = () => useContext(LightboxContext);
