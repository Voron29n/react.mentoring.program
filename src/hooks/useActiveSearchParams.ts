import { useCallback, useEffect, useRef } from 'react';
import { useSearchParams } from 'react-router-dom';
import { IDropDownItem } from 'components';

export const useActiveSearchParams = (
  changeCallback: () => void,
  searchName: string,
  activeList: Array<IDropDownItem>,
  activeValue: IDropDownItem,
  setActiveValue: (
    activeSortType: IDropDownItem,
    searchParams: URLSearchParams,
    setSearchParams: (searchParams: URLSearchParams) => void
  ) => void
) => {
  const [searchParams, setSearchParams] = useSearchParams();
  const refActive = useRef<boolean>(false);

  useEffect(() => {
    const searchParam = searchParams.get(searchName);
    const searchValue = activeList.find(({ value }) => value === searchParam);
    setActiveValue(
      searchValue ? searchValue : activeList[0],
      searchParams,
      setSearchParams
    );
  }, []);

  useEffect(() => {
    if (!refActive.current) {
      refActive.current = true;
      return;
    }

    changeCallback();
  }, [activeValue]);

  const handleSelectedActive: (selectedValue: IDropDownItem) => void =
    useCallback(
      (selectedValue: IDropDownItem) => {
        setActiveValue(selectedValue, searchParams, setSearchParams);
      },
      [setActiveValue]
    );

  return { handleSelectedActive };
};
