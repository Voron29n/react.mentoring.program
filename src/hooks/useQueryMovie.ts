import { useQuery, UseQueryResult } from '@tanstack/react-query';
import { Movie } from 'components';
import {
  fetchReq,
  HttpMethod,
  ServerMovie,
  transformMovieMap
} from 'hooks/utils';
import { baseApiConfig } from 'utils';

const fetchDeleteMovie = async (id: number): Promise<Response> => {
  return fetchReq({
    url: `${baseApiConfig._apiBase}movies/${id}`,
    method: HttpMethod.DELETE
  });
};

const fetchEditMovie = async (movie: Movie): Promise<ServerMovie> => {
  const serverMovie: ServerMovie = transformMovieMap<ServerMovie>(movie);
  const res = await fetchReq({
    url: `${baseApiConfig._apiBase}movies`,
    formData: JSON.stringify(serverMovie),
    method: HttpMethod.PUT,
    headers: {
      'Content-Type': 'application/json'
    }
  });
  return res.json();
};

const fetchAddMovie = async (movie: Movie): Promise<ServerMovie> => {
  const serverMovie: ServerMovie = transformMovieMap<ServerMovie>(movie);
  delete serverMovie.id;
  const res = await fetchReq({
    url: `${baseApiConfig._apiBase}movies`,
    formData: JSON.stringify(serverMovie),
    method: HttpMethod.POST,
    headers: {
      'Content-Type': 'application/json'
    }
  });
  return res.json();
};

const fetchGetMovie = async (movieId: number): Promise<ServerMovie> => {
  const res = await fetchReq({
    url: `${baseApiConfig._apiBase}movies/${movieId}`,
    method: HttpMethod.GET,
    headers: {
      'Content-Type': 'application/json'
    }
  });
  return res.json();
};

export enum MovieQueryTypeActions {
  GET = 'get-movie',
  EDIT = 'edit-movie',
  DELETE = 'delete-movie',
  ADD = 'add-movie'
}

interface IUseQueryMovie {
  movie: Movie;
  isConfirmed: boolean;
  movieTypeAction: MovieQueryTypeActions;
}

export const useQueryMovie = ({
  movie,
  isConfirmed,
  movieTypeAction
}: IUseQueryMovie): UseQueryResult<ServerMovie | Response> => {
  const queryDelete = useQuery({
    queryKey: [MovieQueryTypeActions.DELETE, movie.id],
    queryFn: () => fetchDeleteMovie(movie.id),
    enabled: movieTypeAction == MovieQueryTypeActions.DELETE && isConfirmed
  });
  const queryEdit = useQuery({
    queryKey: [MovieQueryTypeActions.EDIT, movie],
    queryFn: () => fetchEditMovie(movie),
    enabled: movieTypeAction == MovieQueryTypeActions.EDIT && isConfirmed
  });
  const queryAdd = useQuery({
    queryKey: [MovieQueryTypeActions.EDIT, movie],
    queryFn: () => fetchAddMovie(movie),
    enabled: movieTypeAction == MovieQueryTypeActions.ADD && isConfirmed
  });
  const queryGet = useQuery({
    queryKey: [MovieQueryTypeActions.GET, movie.id],
    queryFn: () => fetchGetMovie(movie.id),
    enabled: movieTypeAction == MovieQueryTypeActions.GET && isConfirmed
  });

  switch (movieTypeAction) {
    case MovieQueryTypeActions.GET:
      return queryGet;
    case MovieQueryTypeActions.ADD:
      return queryAdd;
    case MovieQueryTypeActions.EDIT:
      return queryEdit;
    case MovieQueryTypeActions.DELETE:
      return queryDelete;
    default:
      return null;
  }
};
