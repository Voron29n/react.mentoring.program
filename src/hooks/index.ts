import { useActions } from './useActions';
import { useActiveSearchParams } from './useActiveSearchParams';
import { useHovering } from './useHovering';
import { useLightboxContext } from './useLightboxContext';
import { useMovieService } from './useMovieService';
import { MovieQueryTypeActions, useQueryMovie } from './useQueryMovie';
import { useQueryMovieList } from './useQueryMovieList';
import { useTypedDispatch } from './useTypedDispatch';
import { useTypedSelector } from './useTypedSelector';

export {
  useActions,
  useHovering,
  useTypedSelector,
  useTypedDispatch,
  useQueryMovieList,
  useQueryMovie,
  useMovieService,
  useActiveSearchParams,
  useLightboxContext,
  MovieQueryTypeActions
};
