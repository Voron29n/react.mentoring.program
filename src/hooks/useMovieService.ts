import { useCallback, useEffect, useState } from 'react';
import { Dispatch } from 'redux';
import { Movie } from 'components';
import { MovieQueryTypeActions, useQueryMovie } from 'hooks/useQueryMovie';
import { ServerMovie, transformMovieMap } from 'hooks/utils';
import { MovieDetailActions, MovieListActions } from 'store/actionCreators';

interface IUseMovieService {
  defaultMovie: Movie;
  successCallback?: () => void;
  movieAction: (
    movie: Movie,
    searchParams?: URLSearchParams,
    setSearchParams?: (searchParams: URLSearchParams) => void
  ) => (dispatch: Dispatch<MovieListActions | MovieDetailActions>) => void;
  movieQueryTypeActions: MovieQueryTypeActions;
}

export const useMovieService = ({
  defaultMovie = {} as Movie,
  successCallback = () => {},
  movieAction,
  movieQueryTypeActions
}: IUseMovieService) => {
  const [newMovie, setNewMovie] = useState(defaultMovie);
  const [isConfirmed, setIsConfirmed] = useState(false);
  const { isError, isLoading, isSuccess, data } = useQueryMovie({
    movie: newMovie,
    isConfirmed,
    movieTypeAction: movieQueryTypeActions
  });

  const handleConfirm = useCallback((movie: Movie) => {
    setNewMovie(movie);
    setIsConfirmed(true);
  }, []);

  useEffect(() => {
    if (isSuccess) {
      movieAction(
        (data as ServerMovie).id !== undefined
          ? transformMovieMap<Movie>(data as ServerMovie)
          : newMovie
      );
      successCallback();
    }
  }, [isSuccess, data]);

  return { isError, isLoading, isSuccess, isConfirmed, handleConfirm };
};
