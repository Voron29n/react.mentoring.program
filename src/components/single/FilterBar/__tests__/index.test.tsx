import React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { render, RenderResult } from '@testing-library/react';
import { IDropDownItem } from 'components';
import * as hooks from 'hooks';
import { FilterBar } from '../index';

const mockStore = configureMockStore([thunk]);
const store = mockStore({
  searchMovies: {
    activeGenre: null,
    activeSortType: {
      label: 'test',
      value: ''
    } as IDropDownItem
  }
});

describe('FilterBar component with mock hook', () => {
  const handleSelectedActive = jest.fn();
  let filterBarRender: RenderResult;

  beforeEach(() => {
    jest.spyOn(hooks, 'useActiveSearchParams').mockReturnValue({
      handleSelectedActive
    });
    filterBarRender = render(
      <ReduxProvider store={store}>
        <FilterBar />
      </ReduxProvider>
    );
  });

  it('FilterBar snapshot', () => {
    expect(filterBarRender).toMatchSnapshot();
  });
});
