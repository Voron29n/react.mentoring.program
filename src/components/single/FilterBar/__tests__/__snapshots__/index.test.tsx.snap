// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterBar component with mock hook FilterBar snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="filter__bar"
      >
        <div
          class="container"
        >
          <ul
            class="genre__list"
          >
            <li
              class="genre__item upper__text "
            >
              all
            </li>
            <li
              class="genre__item upper__text "
            >
              Fantasy
            </li>
            <li
              class="genre__item upper__text "
            >
              Adventure
            </li>
            <li
              class="genre__item upper__text "
            >
              Science Fiction
            </li>
            <li
              class="genre__item upper__text "
            >
              Drama
            </li>
          </ul>
          <div
            class="filter__bar__line"
          >
            <div
              class="filter__bar__active__line"
            />
          </div>
          <div
            class="sort__bar__container upper__text"
          >
            <div
              class="label"
            >
              sort by
            </div>
            <div>
              <div
                class="select"
              >
                <div
                  class="active__item"
                >
                  test
                </div>
                <img
                  alt="close"
                  class="select__img "
                  src="[object Object]"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="filter__bar"
    >
      <div
        class="container"
      >
        <ul
          class="genre__list"
        >
          <li
            class="genre__item upper__text "
          >
            all
          </li>
          <li
            class="genre__item upper__text "
          >
            Fantasy
          </li>
          <li
            class="genre__item upper__text "
          >
            Adventure
          </li>
          <li
            class="genre__item upper__text "
          >
            Science Fiction
          </li>
          <li
            class="genre__item upper__text "
          >
            Drama
          </li>
        </ul>
        <div
          class="filter__bar__line"
        >
          <div
            class="filter__bar__active__line"
          />
        </div>
        <div
          class="sort__bar__container upper__text"
        >
          <div
            class="label"
          >
            sort by
          </div>
          <div>
            <div
              class="select"
            >
              <div
                class="active__item"
              >
                test
              </div>
              <img
                alt="close"
                class="select__img "
                src="[object Object]"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
